File: nodeset/iterator.go
// Copyright 2019 Grendel Authors. All rights reserved.
//
// This file is part of Grendel.
//
// Grendel is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Grendel is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Grendel. If not, see <https://www.gnu.org/licenses/>.

package nodeset

import (
	"fmt"
	"sort"

	"github.com/segmentio/fasthash/fnv1a"
)

type NodeSetIterator struct {
	nodes   []string
	current int
}

type RangeSetItem struct {
	value   int
	padding int
}

type RangeSetNDIterator struct {
	vects   [][]*RangeSetItem
	seen    map[uint64]struct{}
	current int
}

func (i *NodeSetIterator) Next() bool {
	i.current++

	if i.current < len(i.nodes) {
		return true
	}

	return false
}

func (i *NodeSetIterator) Len() int {
	return len(i.nodes)
}

func (i *NodeSetIterator) Value() string {
	return i.nodes[i.current]
}

func NewRangeSetNDIterator() *RangeSetNDIterator {
	return &RangeSetNDIterator{
		vects:   make([][]*RangeSetItem, 0),
		seen:    make(map[uint64]struct{}),
		current: -1,
	}
}

func (i *RangeSetNDIterator) Next() bool {
	i.current++

	if i.current < len(i.vects) {
		return true
	}

	return false
}

func (i *RangeSetNDIterator) Len() int {
	return len(i.vects)
}

func (i *RangeSetNDIterator) IntValue() []int {
	vals := make([]int, 0, len(i.vects[i.current]))
	for _, v := range i.vects[i.current] {
		vals = append(vals, v.value)
	}
	return vals
}

func (i *RangeSetNDIterator) FormatList() []interface{} {
	vals := make([]interface{}, 0, len(i.vects[i.current]))
	for _, v := range i.vects[i.current] {
		vals = append(vals, fmt.Sprintf("%0*d", v.padding, v.value))
	}
	return vals
}

func (it *RangeSetNDIterator) Sort() {
	sort.SliceStable(it.vects, func(i, j int) bool {
		for x := range it.vects[i] {
			if it.vects[i][x].value != it.vects[j][x].value {
				return it.vects[i][x].value < it.vects[j][x].value
			}
		}
		return false
	})
}

func (it *RangeSetNDIterator) product(result []*RangeSetItem, params ...[]*RangeSetItem) {
	if len(params) == 0 {
		hash := fnv1a.Init64
		for _, i := range result {
			hash = fnv1a.AddUint64(hash, uint64(i.value))
		}

		if _, ok := it.seen[hash]; !ok {
			it.seen[hash] = struct{}{}
			it.vects = append(it.vects, result)
		}

		return
	}

	p, params := params[0], params[1:]
	for i := 0; i < len(p); i++ {
		resultCopy := append([]*RangeSetItem{}, result...)
		it.product(append(resultCopy, p[i]), params...)
	}
}

File: internal/frontend/src/lib/nodeset/iterator.ts
import { RangeSetItem } from "./range-set";

export class NodeSetIterator implements Iterable<string> {
  private current = -1;

  constructor(private readonly nodes: readonly string[]) {}

  next(): boolean {
    this.current++;
    return this.current < this.nodes.length;
  }

  len(): number {
    return this.nodes.length;
  }

  value(): string {
    return this.nodes[this.current];
  }

  stringSlice(): string[] {
    const out: string[] = [];
    while (this.next()) {
      out.push(this.value());
    }
    return out;
  }

  [Symbol.iterator](): Iterator<string> {
    let i = 0;
    const nodes = this.nodes;
    return {
      next(): IteratorResult<string> {
        if (i < nodes.length) {
          return { value: nodes[i++], done: false };
        }
        return { value: undefined, done: true };
      },
    };
  }
}

export class RangeSetNDIterator implements Iterable<RangeSetItem[]> {
  private readonly vects: RangeSetItem[][] = [];
  private readonly seen = new Set<string>();
  private current = -1;

  next(): boolean {
    this.current++;
    return this.current < this.vects.length;
  }

  len(): number {
    return this.vects.length;
  }

  intValue(): number[] {
    return this.vects[this.current].map((v) => v.value);
  }

  formatList(): string[] {
    return this.vects[this.current].map((v) => v.toString());
  }

  values(): readonly RangeSetItem[][] {
    return this.vects;
  }

  sort(): void {
    this.vects.sort((a, b) => {
      for (let x = 0; x < a.length; x++) {
        if (a[x].value !== b[x].value) {
          return a[x].value - b[x].value;
        }
      }
      return 0;
    });
  }

  product(result: RangeSetItem[], ...params: RangeSetItem[][]): void {
    if (params.length === 0) {
      const key = result.map((i) => i.value).join(",");
      if (!this.seen.has(key)) {
        this.seen.add(key);
        this.vects.push(result);
      }
      return;
    }

    const [p, ...rest] = params;
    for (let i = 0; i < p.length; i++) {
      this.product([...result, p[i]], ...rest);
    }
  }

  [Symbol.iterator](): Iterator<RangeSetItem[]> {
    let i = 0;
    const vects = this.vects;
    return {
      next(): IteratorResult<RangeSetItem[]> {
        if (i < vects.length) {
          return { value: vects[i++], done: false };
        }
        return { value: undefined, done: true };
      },
    };
  }
}
